File: Problems 0000-0099/76. Minimum Window Substring/solution.cpp
class Solution {
    public:
        string minWindow(string s, string t) {
            unordered_map<char, int> freq;
            int numCharsLeft = t.length();
            for (char c : t) {
                freq[c]++;
            }
    
            int n = s.length();
            int minLen = n + 1;
            int left = 0;
            int start = 0;
            for (int right = 0; right < n; right++) {
                if (freq[s[right]] > 0) numCharsLeft--;
                if (freq.find(s[right]) != freq.end()) {
                    freq[s[right]]--;
                }
    
                while (numCharsLeft == 0) {
                    int currLen = right - left + 1;
                    if (currLen < minLen) {
                        minLen = currLen;
                        start = left;
                    }
    
                    if (freq.find(s[left]) != freq.end()) {
                        if (freq[s[left]] == 0) numCharsLeft++;
                        freq[s[left]]++;
                    }
    
                    left++;
                }
    
            }
    
            if (minLen == n + 1) return "";
            return s.substr(start, minLen);
        }
    };

File: Problems 0000-0099/76. Minimum Window Substring/solution.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        tfreq = {}
        for c in t:
            tfreq[c] = tfreq.get(c, 0) + 1
        
        numReq = len(tfreq.keys())
        numReqMet = 0
        sfreq = {}
        left = 0
        n = len(s)
        minLen = n + 1
        start = 0
        for right in range(n):
            sfreq[s[right]] = sfreq.get(s[right], 0) + 1
            if sfreq[s[right]] == tfreq.get(s[right], 0):
                numReqMet += 1

            while numReqMet == numReq:
                currLen = right - left + 1
                if currLen < minLen:
                    minLen = currLen
                    start = left
                sfreq[s[left]] -= 1
                if s[left] in tfreq and sfreq[s[left]] < tfreq[s[left]]:
                    numReqMet -= 1
                if sfreq[s[left]] == 0:
                    del sfreq[s[left]]
                left += 1
        
        if minLen == n + 1:
            return ""
        return s[start:start+minLen]